devices/uart-gr716/uart.c

```diff
@@ -0,0 +1,343 @@
+	{ UART2_BASE, UART2_IRQ, UART2_TX, UART2_RX },
+	{ UART3_BASE, UART3_IRQ, UART3_TX, UART3_RX },
+	{ UART4_BASE, UART4_IRQ, UART4_TX, UART4_RX },
+	{ UART5_BASE, UART5_IRQ, UART5_TX, UART5_RX }
+};
+
+
+static inline void uart_rxData(uart_t *uart)
+{
+	char c;
+	/* Keep getting data until rx fifo is not empty */
+	while ((*(uart->base + uart_status) & DATA_READY) != 0) {
+		c = *(uart->base + uart_data) & 0xff;
+		lib_cbufWrite(&uart->cbuffRx, &c, 1);
+	}
+}
+
+
+static inline void uart_txData(uart_t *uart, const void *buff, size_t len)
+{
+	size_t i;
+	const char *c = buff;
+
+	for (i = 0; i < len; i++) {
+		/* Fill until tx fifo is not full */
+		while ((*(uart->base + uart_status) & TX_FIFO_FULL) != 0) {
+		}
+		*(uart->base + uart_data) = c[i];
+	}
+}
+
+
+static int uart_irqHandler(unsigned int n, void *data)
+{
+	uart_t *uart = (uart_t *)data;
+	u32 status = *(uart->base + uart_status);
+
+	if ((status & DATA_READY) != 0) {
+		uart_rxData(uart);
+	}
+
+	return 0;
+}
+
+
+static int uart_setPin(u8 pin)
+{
+	io_cfg_t uartCfg;
+	uartCfg.opt = 0x1;
+	uartCfg.pin = pin;
+	uartCfg.pullup = 0;
+	uartCfg.pulldn = 0;
+
+	switch (pin) {
+		case UART0_TX:
+		case UART1_TX:
+			/* pins used for SRAM */
+			return 0;
+		case UART2_TX:
+		case UART3_TX:
+		case UART4_TX:
+		case UART5_TX:
+			uartCfg.dir = GPIO_DIR_OUT;
+			break;
+		case UART0_RX:
+		case UART1_RX:
+			/* pins used for SRAM */
+			return 0;
+		case UART2_RX:
+		case UART3_RX:
+		case UART4_RX:
+		case UART5_RX:
+			uartCfg.dir = GPIO_DIR_IN;
+			break;
+		default:
+			return -1;
+	}
+
+	return _gr716_ioCfg(&uartCfg);
+}
+
+
+/* From datasheet:
+ * appropriate formula to calculate the scaler for desired baudrate,
+ * using integer division where the remainder is discarded:
+ * scaler = (sysclk_freq)/(baudrate * 8 + 7)
+ */
+static u32 uart_calcScaler(u32 baud)
+{
+	u32 scaler = (SYSCLK_FREQ / (baud * 8 + 7));
+
+	return scaler;
+}
+
+/* Device interface */
+
+static ssize_t uart_read(unsigned int minor, addr_t offs, void *buff, size_t len, time_t timeout)
+{
+	ssize_t res;
+	uart_t *uart;
+	time_t start;
+
+	if (minor >= UART_MAX_CNT) {
+		return -EINVAL;
+	}
+
+	uart = &uart_common.uarts[minor];
+	start = hal_timerGet();
+	while (lib_cbufEmpty(&uart->cbuffRx) != 0) {
+		if (hal_timerGet() - start > timeout) {
+			return -ETIME;
+		}
+	}
+	hal_interruptsDisableAll();
+	res = lib_cbufRead(&uart->cbuffRx, buff, len);
+	hal_interruptsEnableAll();
+
+	return res;
+}
+
+
+static ssize_t uart_write(unsigned int minor, const void *buff, size_t len)
+{
+	uart_t *uart;
+
+	if (minor >= UART_MAX_CNT) {
+		return -EINVAL;
+	}
+
+	uart = &uart_common.uarts[minor];
+
+	uart_txData(uart, buff, len);
+
+	return len;
+}
+
+
+static ssize_t uart_safeWrite(unsigned int minor, addr_t offs, const void *buff, size_t len)
+{
+	ssize_t res = 0;
+	size_t l;
+
+	for (l = 0; l < len; l += res) {
+		res = uart_write(minor, buff + l, len - l);
+		if (res < 0) {
+			return -ENXIO;
+		}
+	}
```

Comment:
Do not use `l` as a variable name, for isn't the best choice here imho
```suggestion
	size_t wrote = 0;

	while (wrote < len) {
		res = uart_write(minor, buff + l, len - l);
		if (res < 0) {
			return -ENXIO;
		}
		wrote += res;
	}
```